from os.path import join as pjoin
configfile: "config.yml"


ref = config["ref"]
variants = config["vcf"]
rsample = config["samples"]["ref"]
qsample = config["samples"]["query"]
rcontigs = config["contigs"]["ref"]
qcontigs = config["contigs"]["query"]
outdir = config["outdir"]
nths = config["threads"]

sdir = pjoin(config["codedir"], "experiments", "realdata", "scripts")
stellabin = pjoin(config["codedir"], "PingPong")

# all_specific = pjoin(outdir, "pp", "sspecific.fq")
# short_specific = pjoin(outdir, "pp", "sspecific.short.fq")
# # mid_specific = pjoin(outdir, "pp", "sspecific.mid.fq")
# long_specific = pjoin(outdir, "pp", "sspecific.long.fq")
#
# all_specific_ov = pjoin(outdir, "pp-overlap", "sspecific.fq")
# short_specific_ov = pjoin(outdir, "pp-overlap", "sspecific.short.fq")
# # mid_specific_ov = pjoin(outdir, "pp-overlap", "sspecific.mid.fq")
# long_specific_ov = pjoin(outdir, "pp-overlap", "sspecific.long.fq")



rule run:
    input:
        pjoin(outdir, "pp", "over", "4paneplot.png"),
        pjoin(outdir, "pp", "nonover", "4paneplot.png"),
        ### contig-based analysis ###
        pjoin(outdir, "pp", "nonover", "Cprecision.txt"),
        #pjoin(outdir, "pp", "nonover", "Cplot.png"), # already in 4paneplot
        pjoin(outdir, "pp", "over", "Cprecision.txt"),
        # pjoin(outdir, "pp", "over", "Cplot.png"), # already in 4paneplot
        pjoin(outdir, "kmers", "31mers", "Cprecision.txt"),
        pjoin(outdir, "kmers", "101mers", "Cprecision.txt"),
        pjoin(outdir, "kmers", "31tigs", "Cprecision.txt"),
        # pjoin(outdir, "kmers", "101tigs", "Cprecision.txt"), # minimap2: format.c:378: write_sam_cigar: Assertion clip_len[0] < qlen && clip_len[1] < qlen' failed.
        ### haplotype-based analysis ###
        expand(pjoin(outdir, "pp", "nonover", "precision.{subset}.txt"),
               subset=["short", "long", "all"]),
        expand(pjoin(outdir, "pp", "nonover", "recall.{subset}.txt"),
               subset=["short", "long", "all"]),
        # pjoin(outdir, "pp", "nonover", "nonhaploplot.png"), # already in 4paneplot
        expand(pjoin(outdir, "pp", "over", "precision.{subset}.txt"),
               subset=["short", "long", "all"]),
        expand(pjoin(outdir, "pp", "over", "recall.{subset}.txt"),
               subset=["short", "long", "all"]),
        # pjoin(outdir, "pp", "over", "nonhaploplot.png"), # already in 4paneplot
        ### kmers ### # rules for alignment are hardcoded to 31 and 101 only
        pjoin(outdir, "kmers", "31mers", "recall.txt"),
        pjoin(outdir, "kmers", "31mers", "precision.txt"),
        pjoin(outdir, "kmers", "101mers", "recall.txt"),
        pjoin(outdir, "kmers", "101mers", "precision.txt"),

        expand(pjoin(outdir, "kmers", "31tigs", "precision.{subset}.txt"),
               subset = ["short", "long", "all"]),
        expand(pjoin(outdir, "kmers", "31tigs", "recall.{subset}.txt"),
               subset = ["short", "long", "all"]),
        expand(pjoin(outdir, "kmers", "101tigs", "precision.{subset}.txt"),
               subset = ["short", "long", "all"]),
        expand(pjoin(outdir, "kmers", "101tigs", "recall.{subset}.txt"),
               subset = ["short", "long", "all"]),

        # some other plots
        pjoin(outdir, "samples.hist.png"),
        pjoin(outdir, "pp", "nonover", "correlation.png"),
        pjoin(outdir, "pp", "over", "correlation.png"),


include: "haplopre.sm"
include: "contigspre.sm"
include: "pp.sm"
include: "ppcontigs.sm"
include: "pphaplo.sm"
include: "kmers.sm"
include: "kmerscontigs.sm"
include: "kmershaplo.sm"
include: "unitigs.sm"
include: "unitigscontigs.sm"
include: "unitigshaplo.sm"
include: "utils.sm"
